apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,228 @@
+	// name of this flavor. The name should match the .metadata.name of a
+	// ResourceFlavor. If a matching ResourceFlavor does not exist, the
+	// ClusterQueue will have an Active condition set to False.
+	Name ResourceFlavorReference `json:"name"`
+
+	// resources is the list of quotas for this flavor per resource.
+	// There could be up to 16 resources.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Resources []ResourceQuota `json:"resources"`
+}
+
+type ResourceQuota struct {
+	// name of this resource.
+	Name corev1.ResourceName `json:"name"`
+
+	// quota is the quantity of this resource that is available for Workloads
+	// admitted by this ClusterQueue at a point in time.
+	// The quota must non-negative.
+	//
+	// If the ClusterQueue belongs to a cohort, the sum of the quotas for each
+	// (flavor, resource) combination defines the maximum quantity that can be
+	// allocated by a ClusterQueue in the cohort.
+	Quota resource.Quantity `json:"quota"`
+
+	// borrowingLimit is the maximum amount of quota for the (flavor, resource)
+	// combination that this ClusterQueue is allowed to borrow from the unused
+	// quota of other ClusterQueues in the same cohort.
+	// If null, it means that there is no borrowing limit.
+	// If not null, it must be non-negative.
+	// borrowingLimit must be null if spec.cohort is empty.
+	// +optional
+	BorrowingLimit *resource.Quantity `json:"borrowingLimit,omitempty"`
+}
+
+// ResourceFlavorReference is the name of the ResourceFlavor.
+type ResourceFlavorReference string
+
+// ClusterQueueStatus defines the observed state of ClusterQueue
+type ClusterQueueStatus struct {
+	// flavorUsage are the used quotas by flavor currently in use by the workloads
+	// assigned to this ClusterQueue.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	// +optional
+	FlavorUsage []FlavorUsage `json:"flavorUsage"`
+
+	// PendingWorkloads is the number of workloads currently waiting to be
+	// admitted to this clusterQueue.
+	// +optional
+	PendingWorkloads int32 `json:"pendingWorkloads"`
+
+	// AdmittedWorkloads is the number of workloads currently admitted to this
+	// clusterQueue and haven't finished yet.
+	// +optional
+	AdmittedWorkloads int32 `json:"admittedWorkloads"`
+
+	// conditions hold the latest available observations of the ClusterQueue
+	// current state.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type FlavorUsage struct {
+	// name of the flavor.
+	Name string `json:"name"`
+
+	// resources lists the quota usage for the resources in this flavor.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Resources []ResourceUsage `json:"resources"`
+}
+
+type ResourceUsage struct {
+	// name of the resource
+	Name corev1.ResourceName `json:"name"`
+
+	// Total is the total quantity of the resource used, including resources
```

Comment:
We cannot quantify those. A ClusterQueue doesn't directly borrow from another ClusterQueue. They borrow from the pool of unused resources in the Cohort.

This is so that a single Workload can use borrowed quota from multiple CQs at the same time.

Reply:
Well, we kind-of do.  Imagine the situation when all capacity in the cohort is taken and the queue itself has no admitted workloads. Then borrowedByCohort = base queue quota. If All queues have low usage, within their quota then borrowedByCohort = 0. Right :)? So we can generalize it to:  
``` 
max(0, SumOfUsageInOtherQueuesInCohort - SumOfQuotaInOtherQuesesInCohort)
```

Reply:
Good point, but it can be added later. Not sure how much worth it is. We don't maintain those numbers in the cache. We just calculate them when we snapshot the cache to run the admission cycle.